extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheResultInterceptor.java

```diff
@@ -42,43 +42,69 @@ public Object intercept(InvocationContext invocationContext) throws Throwable {
         }
 
         try {
-
-            Uni<Object> cacheValue = cache.get(key, new Function<Object, Object>() {
-                @Override
-                public Object apply(Object k) {
-                    try {
-                        if (Uni.class.isAssignableFrom(invocationContext.getMethod().getReturnType())) {
+            final boolean isUni = Uni.class.isAssignableFrom(invocationContext.getMethod().getReturnType());
+            if (isUni) {
+                Uni<Object> ret = cache.get(key, new Function<Object, Object>() {
+                    @Override
+                    public Object apply(Object k) {
+                        try {
                             LOGGER.debugf("Adding %s entry with key [%s] into cache [%s]",
                                     UnresolvedUniValue.class.getSimpleName(), key, binding.cacheName());
                             return UnresolvedUniValue.INSTANCE;
-                        } else {
-                            return invocationContext.proceed();
+                        } catch (Throwable e) {
+                            throw new CacheException(e);
+                        }
+                    }
+                }).onItem().transformToUni(o -> {
+                    if (o == UnresolvedUniValue.INSTANCE) {
+                        try {
+                            return ((Uni<Object>) invocationContext.proceed())
+                                    .onItem().call(emittedValue -> cache.replaceUniValue(key, emittedValue));
+                        } catch (Exception e) {
+                            throw new CacheException(e);
                         }
-                    } catch (Throwable e) {
-                        throw new CacheException(e);
+                    } else {
+                        return Uni.createFrom().item(o);
                     }
+                });
+                if (binding.lockTimeout() <= 0) {
+                    return ret;
                 }
-            });
+                return ret.ifNoItem().after(Duration.ofMillis(binding.lockTimeout())).recoverWithUni(() -> {
+                    try {
+                        return (Uni<?>) invocationContext.proceed();
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
```

Comment:
```suggestion
                        throw new CacheException(e);
```

Reply:
I'd also suggest rethrowing `CacheException` by adding another catch block instead of rewrapping it here